pkg/cloud/services/networking/port_test.go

```diff
@@ -0,0 +1,373 @@
+							Description:         "Created by cluster-api-provider-openstack cluster test-cluster",
+							SecurityGroups:      &portSecurityGroups,
+							NetworkID:           netID,
+							AllowedAddressPairs: []ports.AddressPair{},
+						},
+					},
+					).Return(&ports.Port{ID: portID1}, nil)
+			},
+			&ports.Port{ID: portID1},
+			false,
+		},
+		{
+			"creates port with tags passed to function",
+			"foo-port-1",
+			infrav1.Network{
+				ID:       netID,
+				PortOpts: &infrav1.PortOpts{},
+			},
+			nil,
+			[]string{"my-tag"},
+			func(m *mock_networking.MockNetworkClientMockRecorder) {
+				// No ports found
+				m.
+					ListPort(ports.ListOpts{
+						Name:      "foo-port-1",
+						NetworkID: netID,
+					}).Return([]ports.Port{}, nil)
+				m.CreatePort(portsbinding.CreateOptsExt{
+					CreateOptsBuilder: ports.CreateOpts{
+						Name:                "foo-port-1",
+						Description:         "Created by cluster-api-provider-openstack cluster test-cluster",
+						NetworkID:           netID,
+						AllowedAddressPairs: []ports.AddressPair{},
+					},
+				}).Return(&ports.Port{ID: portID1}, nil)
+				m.ReplaceAllAttributesTags(0, 0, 0)
+			},
+			&ports.Port{ID: portID1},
+			false,
+		},
+		{
+			"creates port and trunk (with tags) if they aren't found",
+			"foo-port-1",
+			infrav1.Network{
+				ID: netID,
+				PortOpts: &infrav1.PortOpts{
+					Trunk: pointerToTrue,
+				},
+			},
+			nil,
+			[]string{"my-tag"},
+			func(m *mock_networking.MockNetworkClientMockRecorder) {
+				// No ports found
+				m.
+					ListPort(ports.ListOpts{
+						Name:      "foo-port-1",
+						NetworkID: netID,
+					}).Return([]ports.Port{}, nil)
+				m.
+					CreatePort(portsbinding.CreateOptsExt{
+						CreateOptsBuilder: ports.CreateOpts{
+							Name:                "foo-port-1",
+							Description:         "Created by cluster-api-provider-openstack cluster test-cluster",
+							NetworkID:           netID,
+							AllowedAddressPairs: []ports.AddressPair{},
+						},
+					}).Return(&ports.Port{Name: "foo-port-1", ID: portID1}, nil)
+				m.
+					ListTrunk(trunks.ListOpts{
+						Name:   "foo-port-1",
+						PortID: portID1,
+					}).Return([]trunks.Trunk{}, nil)
+				m.
+					CreateTrunk(trunks.CreateOpts{
+						Name:        "foo-port-1",
+						PortID:      portID1,
+						Description: "Created by cluster-api-provider-openstack cluster test-cluster",
+					}).Return(&trunks.Trunk{ID: trunkID}, nil)
+				// And set tags on the trunk
+				m.ReplaceAllAttributesTags("trunks", trunkID, attributestags.ReplaceAllOpts{Tags: []string{"my-tag"}})
+			},
+			&ports.Port{Name: "foo-port-1", ID: portID1},
+			false,
+		},
+	}
+
+	eventObject := &infrav1.OpenStackMachine{}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.name == "creates port with tags passed to function" {
+				t.Skip("TODO: Bug! Tags aren't applied to the port. See #1016. Then ensure that the port tags are applied.")
```

Comment:
Note that I've skipped a test! This depends on #1016 being merged and then hopefully we will avoid this regression again in future.